File: sum-of-root-to-leaf-binary-numbers.ts
{
    function sumRootToLeaf(root: TreeNode | null): number {
        const sums: string[] = [];

        function dfs(node: TreeNode | null, digits: string) {
            if (!node)
                return;

            digits += node.val;

            if (!node.left && !node.right) {
                sums.push(digits);
                return;
            }

            dfs(node.left, digits);
            dfs(node.right, digits);
        }

        dfs(root, "");

        return sums.reduce((sum, current) => {
            return sum + parseInt(current, 2);
        }, 0)
    };

    class TreeNode {
        val: number
        left: TreeNode | null
        right: TreeNode | null
        constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
            this.val = (val===undefined ? 0 : val)
            this.left = (left===undefined ? null : left)
            this.right = (right===undefined ? null : right)
        }
    }

    const test = new TreeNode(
        1,
        new TreeNode(0, new TreeNode(0), new TreeNode(1)),
        new TreeNode(1, new TreeNode(0), new TreeNode(1)),
    )

    console.log(sumRootToLeaf(test))
}

File: sum-of-root-to-leaf-binary-numbers.rs
use std::cell::RefCell;
use std::rc::Rc;
type OptNode = Option<Rc<RefCell<TreeNode>>>;

pub fn sum_root_to_leaf(root: OptNode) -> i32 {
    Self::find_sum(&root, 0)
}

fn find_sum(node: &OptNode, acc: i32) -> i32 {
    match node.as_ref() {
        None => 0,
        Some(n) => {
            let b = n.borrow();
            let acc2 = (acc << 1) | b.val;
            let l_r = Self::find_sum(&b.left, acc2) + Self::find_sum(&b.right, acc2);
            if l_r == 0 {
                acc2
            } else {
                l_r
            }
        }
    }
}
